pkg/buildchange/change_processor.go

```diff
@@ -0,0 +1,111 @@
+package buildchange
+
+import (
+	"encoding/json"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/pivotal/kpack/pkg/apis/build/v1alpha1"
+)
+
+const (
+	reasonsSeparator = ","
+	errorSeparator   = "\n"
+)
+
+func NewChangeProcessor() *ChangeProcessor {
+	return &ChangeProcessor{
+		changes: []GenericChange{},
+		errStrs: []string{},
+	}
+}
+
+type ChangeProcessor struct {
+	changes []GenericChange
+	errStrs []string
+}
+
+func (c *ChangeProcessor) Process(change Change) *ChangeProcessor {
+	if change == nil {
+		return c
+	}
+
+	if !change.Reason().IsValid() {
+		errStr := fmt.Sprintf("unsupported change reason '%s'", change.Reason())
+		c.errStrs = append(c.errStrs, errStr)
+		return c
+	}
+
+	valid, err := change.IsValid()
+	if err != nil {
+		err := errors.Wrapf(err, "error validating change for reason '%s'", change.Reason())
+		c.errStrs = append(c.errStrs, err.Error())
+	} else if valid {
+		c.changes = append(c.changes, newGenericChange(change))
+	}
+
+	return c
+}
+
+func (c *ChangeProcessor) Summarize() (ChangeSummary, error) {
+	changesStr, err := c.changesStr()
+	if err != nil {
+		err := errors.Wrapf(err, "error generating changes string")
+		c.errStrs = append(c.errStrs, err.Error())
+	}
+
+	summary, err := NewChangeSummary(c.hasChanges(), c.reasonsStr(), changesStr)
+	if err != nil {
+		err := errors.Wrapf(err, "error summarizing changes")
+		c.errStrs = append(c.errStrs, err.Error())
+	}
+
+	if len(c.errStrs) > 0 {
+		return summary, errors.New(strings.Join(c.errStrs, errorSeparator))
+	}
+
+	return summary, nil
+}
+
+func (c *ChangeProcessor) hasChanges() bool {
+	return len(c.changes) > 0
+}
+
+func (c *ChangeProcessor) reasonsStr() string {
+	if !c.hasChanges() {
+		return ""
+	}
+
+	var reasons = make([]string, len(c.changes))
+	for i, change := range c.changes {
+		reasons[i] = change.Reason
+	}
+
+	sort.SliceStable(reasons, func(i, j int) bool {
+		return strings.Index(v1alpha1.BuildReasonSortIndex, reasons[i]) <
+			strings.Index(v1alpha1.BuildReasonSortIndex, reasons[j])
+	})
+
+	return strings.Join(reasons, reasonsSeparator)
+}
+
+func (c *ChangeProcessor) changesStr() (string, error) {
+	if !c.hasChanges() {
+		return "", nil
+	}
+
+	sort.SliceStable(c.changes, func(i, j int) bool {
```

Comment:
Is this sorting necessary?

Reply:
> Is this sorting necessary?

The sorting is done to ensure that the reasons string and the corresponding changes are listed in the same order.
A safety, to ensure that the Process() change call sequence on the ChangeProcessor does not determine the output order.
I think it is a small overhead to ensure changes on the build_needed.go file does not inadvertently change the order for printing changes.

Reply:
@matthewmcnew please let me know if you still consider this as unnecessary. I would prefer to keep it. Either way, would like to resolve this comment.

Reply:
Similar to the [validation discussion](https://github.com/pivotal/kpack/pull/556/files#r538868069) it seems like the change processor shouldn't need to understand what types it is processing. It seems like the order of the calls to `Process()` should be sufficent to enforce order.

Reply:
I do not agree with your observation and am hesitant to make change which I am not convinced is correct.  Maybe we should chat about this?

The ChangeProcess processes Changes for what? Producing the Reasons and Changes string. It is the output of the Summarize call on the ChangeProcessor. Considering that, I disagree with your premise that the ChangeProcessor should not be responsible for ensuring the reasons and changes string are sorted the same way.

Reply:
> It seems like the order of the calls to Process() should be sufficent to enforce order.

I did address this in my [prior comment](https://github.com/pivotal/kpack/pull/556#discussion_r538882526)

At least 2 ways that I can see foresee change order going out of sync with reasons (by oversight)
- BuildReasonSortIndex is updated without a similar update to the Process() call sequence
- Future PR/Code changes which involve a reorder of the Process() call sequence without an update to the BuildReasonSortIndex

I feel an implicit dependence on the Process() call sequence (vs the current implementation of explicit ordering of reasons and changes) makes it very fragile to unintended breakage - i.e. the reasons listing and changes listing go out of sequence.

Reply:
Update from discussion: `Is this sorting necessary?` was addressing both the sorting of changes and reasons within ChangeProcessor; which would mean the possibility for reasons string and change string going out of sync is removed.
Moving ahead with dropping sort of both changes and reason in change_processor

Reply:
addressed in https://github.com/pivotal/kpack/commit/e73935fa92379241fdd65e1d2d916a4ae487afc5